bson/src/main/org/bson/codecs/FlexibleStringCodec.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs;
+
+import org.bson.BsonInvalidOperationException;
+import org.bson.BsonReader;
+import org.bson.BsonType;
+import org.bson.BsonWriter;
+import org.bson.codecs.configuration.CodecConfigurationException;
+import org.bson.types.ObjectId;
+
+/**
+ * A codec capable of decoding different types to Strings and encoding
+ * Strings to different types depending on the provided BsonRepresentation.
+ *
+ * @since 4.2
+ */
+public class FlexibleStringCodec extends StringCodec implements FlexibleCodec<String> {
```

Comment:
I'd like to consider a different design for handling `BsonRepresentation` that works within the existing codec registry system.  Instead of creating a new codec, and a new provider, what if we made `StringCodec` implement `FlexibleCodec` itself?  That way, the existing codec for string is responsible for determining whether it can handle alternate representations.  This will allow application-defined codecs to participate in this system, and make it easier to add support to more of the built-in codecs.

To make this work though, we have to preserve immutability of codec instances.  So I imagine the interface looking something like this:

    public interface RepresentationConfigurable<T> {
         // get the current representation
         BsonType getRepresentation();

         // Return a codec with the given representation
         // Be careful to preserve immutability!
         // Throw an exception if the requested representation is not supported
         Codec<T> withRepresentation(BsonType representation)
    }

The POJO codec will do an `instanceof RepresentationConfigurable` on the codec for the field with a BsonRepresentation annotation to determine whether the field can participate in this scheme.

Let me know what you think of this alternative.

Reply:
Yup this design sounds great to me. I had considered doing something like this but I wasn't sure if doing the `instanceof` check was considered bad. I hadn't thought of the fact that it will allow application-defined codecs to participate, that's a good point. It'll also just make the code simpler and more concise, so I'll get started on changing it.